### PR TITLE
New Check: Certificate Expiration

### PR DESCRIPTION
**Pull request by LowlyDBA**
Adding a check for the expiration of TDE certificates in databases.

I'm unsure the policy config names should be under policy.database.xxx or remain as is, since the database exclusion is at the server level, but the check itself is at the database level...

Database exclusion for this check is a little kludgy, but was having trouble getting more straightforward approaches to work properly for some reason.

### CODE REVIEW

**Comment by potatoqualitee** [comment]
wow super nice! thank you!

**Comment by michalporeba** [comment]
A very good idea for a test.

A suggestion: so far everywhere I have seen in the codebase double quotes are used for strings. I think it would be worth keeping it consistent and not switch between single and double quotes.

**Comment by LowlyDBA** [comment]
@michalporeba  Good point. After a quick glance I see they're a bit mixed. Happy to standardize on whichever is preferred, though.

File: checks/Database.Tests.ps1 [hunk 1]
```diff
@@ -18,7 +18,7 @@ Describe "Suspect Page" -Tags SuspectPage, $filename {
             @(Get-DbaDatabase -SqlInstance $psitem).ForEach{
                 $results = Get-DbaSuspectPage -SqlInstance $psitem.Parent -Database $psitem.Name
                 It "$psitem should return 0 suspect pages on $($psitem.SqlInstance)" {
-                    @($results).Count | Should -Be 0 -Because 'You dont want suspect pages'
+                    @($results).Count | Should -Be 0 -Because 'You do not want suspect pages'
```
Comment (SQLDBAWithABeard) [inline checks/Database.Tests.ps1]: I think we should standardise to double quotes

File: checks/Database.Tests.ps1 [hunk 2]
```diff
@@ -130,11 +130,12 @@ Describe "Column Identity Usage" -Tags IdentityUsage, $filename {
 
 Describe "Recovery Model" -Tags RecoveryModel, DISA, $filename {
     (Get-SqlInstance).ForEach{
```
Comment (SQLDBAWithABeard) [inline checks/Database.Tests.ps1]: Please use @(Get-SqlInstance).ForEach{ for PS v4 compatability

File: checks/Database.Tests.ps1 [hunk 3]
```diff
@@ -130,11 +130,12 @@ Describe "Column Identity Usage" -Tags IdentityUsage, $filename {
...
+    $recoverymodel = Get-DbcConfigValue policy.recoverymodel.type
         Context "Testing Recovery Model on $psitem" {
             $exclude = Get-DbcConfigValue policy.recoverymodel.excludedb
             @(Get-DbaDbRecoveryModel -SqlInstance $psitem -ExcludeDatabase $exclude).ForEach{
-                It "$($psitem.Name) Should -Be set to $((Get-DbcConfigValue policy.recoverymodel.type)) on $($psitem.SqlInstance)" {
-                    $psitem.RecoveryModel | Should -Be (Get-DbcConfigValue policy.recoverymodel.type) -Because 'You expect this recovery model'
+                It "$($psitem.Name) Should -Be set to $recoverymodel on $($psitem.SqlInstance)" {
```
Comment (SQLDBAWithABeard) [inline checks/Database.Tests.ps1]: Should Be  not Should -Be in title :-) Thats my favourite trick as well!!

File: checks/Database.Tests.ps1 [hunk 4]
```diff
@@ -382,6 +382,23 @@ Describe "Correctly sized Filegroup members" -Tags FileGroupBalanced, $filename
     }
 }
 
+Describe "Certificate Expiration" -Tags CertificateExpiration, $filename {
+    $CertificateWarning = Get-DbcConfigValue policy.certificateexpiration.warningwindow
+    $exclude = Get-DbcConfigValue policy.certificateexpiration.excludedb
+    (Get-SqlInstance).ForEach{
```
Comment (SQLDBAWithABeard) [inline checks/Database.Tests.ps1]: Please use @(Get-SqlInstance).ForEach{ for PS v4 compatability

File: checks/Database.Tests.ps1 [hunk 5]
```diff
@@ -382,6 +382,23 @@ Describe "Correctly sized Filegroup members" -Tags FileGroupBalanced, $filename
...
+        Context "Checking that encryption certificates have not expired on $psitem" {
+            (Get-DbaDatabaseEncryption -SqlInstance $psitem -IncludeSystemDBs | Where-Object {$_.Encryption -eq 'Certificate' -and !($exclude.contains($_.Database))}).ForEach{
```
Comment (SQLDBAWithABeard) [inline checks/Database.Tests.ps1]: Please use @().ForEach{ for PS v4 compatability

File: checks/Database.Tests.ps1 [hunk 6]
```diff
@@ -382,6 +382,23 @@ Describe "Correctly sized Filegroup members" -Tags FileGroupBalanced, $filename
...
+                It "$($psitem.Name) in $($psitem.Database) has not expired" {
+                    $psitem.ExpirationDate  | Should -BeGreaterThan (Get-Date) -Because 'this certificate should not be expired'
+                }
+                It "$($psitem.Name) in $($psitem.Database) does not expire for more than $CertificateWarning months" {
+                    $psitem.ExpirationDate  | Should -BeGreaterThan (Get-Date).AddMonths($CertificateWarning) -Because 'this certificate will soon expire'
```
Comment (SQLDBAWithABeard) [inline checks/Database.Tests.ps1]: Suggest changing soon expire to "expires inside the warning window of $CertificateWarning

**Comment by SQLDBAWithABeard** [comment]
Thank you @LowlyDBA

**Pull request merged**